### PR TITLE
OpAtomicMaxMin, RemoveUnusedMemory, OpSelect

### PR DESCRIPTION
**Pull request by tonghaining**

### CODE REVIEW

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveUnusedMemory.java [hunk 1]
```diff
@@ -30,9 +32,39 @@ public void run(Program program) {
...
+    }
+
+    private void addSpecMemoryObjects(AbstractAssert spec, MemoryObjectCollector collector) {
+        if (spec instanceof AssertInline assertInline) {
+            addExprMemoryObject(assertInline.getAssertion().getExpression(), collector);
+        }
+        if (spec instanceof AssertBasic assertBasic) {
+            addExprMemoryObject(assertBasic.getLeft(), collector);
+            addExprMemoryObject(assertBasic.getRight(), collector);
+        }
```
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveUnusedMemory.java:47]: you can just do `expr.accept(collector)` in both cases here and the `addExprMemoryObject` method becomes obsolete.
However, you also need to overwrite `MemoryObjectCollector.visitLocation` (which could just visit the underlying memory object).

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveUnusedMemory.java [hunk 2]
```diff
@@ -30,9 +32,39 @@ public void run(Program program) {
...
+    }
+
+    private void addExprMemoryObject(Expression expr, MemoryObjectCollector collector) {
+        if (expr instanceof Location loc) {
+            loc.getMemoryObject().accept(collector);
+        }
+        if (expr instanceof Register reg) {
+            reg.getFunction().accept(collector);
+        }
```
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveUnusedMemory.java:67]: I'm not sure what this was supposed to do, but I think this does not what you intended it to do :).

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveUnusedMemory.java (hunk 1 above)
Comment (tonghaining) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveUnusedMemory.java:47]: Thanks!

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveUnusedMemory.java (hunk 2 above)
Comment (tonghaining) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveUnusedMemory.java:67]: Ahh.. I forgot to delete it

File: dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsAtomic.java [hunk 3]
```diff
@@ -115,6 +118,59 @@ public Event visitOpAtomicXor(SpirvParser.OpAtomicXorContext ctx) {
...
+        String scope = builder.getScope(scopeCtx.getText());
+        Set<String> tags = builder.getSemantics(tagsCtx.getText());
+        tags.add(builder.getExpressionStorageClass(ptrCtx.getText()));
+        Expression originalValue = ExpressionFactory.getInstance().makeIntegerCast(ptr, TYPE_FACTORY.getArchType(), true);
+        Expression newValue = ExpressionFactory.getInstance().makeIntegerCast(value, TYPE_FACTORY.getArchType(), true);
+        Expression cmpExpr = ExpressionFactory.getInstance().makeIntCmp(originalValue, kind, newValue);
+        Expression booleanExpr = ExpressionFactory.getInstance().makeBooleanCast(cmpExpr);
+        Expression ite =  ExpressionFactory.getInstance().makeITE(booleanExpr, originalValue, newValue);
+        Expression ite_casted = ExpressionFactory.getInstance().makeIntegerCast(ite, (IntegerType) ptr.getType(), true);
+        SpirvXchg xchg = newSpirvXchg(register, ptr, ite_casted, scope, tags);
```
Comment (natgavrilenko) [inline dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsAtomic.java:170]: I think you will need a new event SpirvRmwExtremum, and add a new compilation for it in VisitorSpirvVulkan. Then ITE should be not on the pointer but on the result register of the load. I.e. now we have

bv64 DUMMY_REG_2 = load(*&%var2, ACQ)
rmw store(*&%var2, ITE((&%var2 > bv64(-3)), &%var2, bv64(-3)), REL)

but it should be 

bv64 DUMMY_REG_2 = load(*&%var2, ACQ)
rmw store(*&%var2, ITE((DUMMY_REG_2 > bv64(-3)), DUMMY_REG_2, bv64(-3)), REL)


I would also add an end-to-end test to make sure it works correctly, for example

; @Input:  %out_result = {0, 0, 0, 0}
; @Output: forall (%out_result[0]==2 and %out_result[1]==2 and %out_result[2]==2 and %out_result[3]==2)
; @Output: forall (%out_final[0]==-3 and %out_final[1]==2 and %out_final[2]==2 and %out_final[2]==-3)
; @Config: 1, 1, 1
; SPIR-V
; Version: 1.0
; Schema: 0
               OpCapability Shader
        %ext = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint GLCompute %main "main"
               OpSource GLSL 450
       %void = OpTypeVoid
       %bool = OpTypeBool
       %func = OpTypeFunction %void
        %int = OpTypeInt 64 1
      %v4int = OpTypeVector %int 4
    %ptr_int = OpTypePointer Uniform %int
  %ptr_v4int = OpTypePointer Output %v4int
     %c3_neg = OpConstant %int -3
         %c0 = OpConstant %int 0
         %c1 = OpConstant %int 1
         %c2 = OpConstant %int 2
         %c3 = OpConstant %int 3
        %c64 = OpConstant %int 64
       %var0 = OpVariable %ptr_int Uniform %c2
       %var1 = OpVariable %ptr_int Uniform %c2
       %var2 = OpVariable %ptr_int Uniform %c2
       %var3 = OpVariable %ptr_int Uniform %c2
 %out_result = OpVariable %ptr_v4int Output
  %out_final = OpVariable %ptr_v4int Output
       %main = OpFunction %void None %func
      %label = OpLabel

    %result0 = OpAtomicSMin %int %var0 %c1 %c64 %c3_neg
    %result1 = OpAtomicUMin %int %var1 %c1 %c64 %c3_neg
    %result2 = OpAtomicSMax %int %var2 %c1 %c64 %c3_neg
    %result3 = OpAtomicUMax %int %var3 %c1 %c64 %c3_neg

     %final0 = OpLoad %int %var0
     %final1 = OpLoad %int %var1
     %final2 = OpLoad %int %var2
     %final3 = OpLoad %int %var3

  %ptr_res_0 = OpAccessChain %ptr_int %out_result %c0
  %ptr_res_1 = OpAccessChain %ptr_int %out_result %c1
  %ptr_res_2 = OpAccessChain %ptr_int %out_result %c2
  %ptr_res_3 = OpAccessChain %ptr_int %out_result %c3

               OpStore %ptr_res_0 %result0
               OpStore %ptr_res_1 %result1
               OpStore %ptr_res_2 %result2
               OpStore %ptr_res_3 %result3

  %ptr_fin_0 = OpAccessChain %ptr_int %out_final %c0
  %ptr_fin_1 = OpAccessChain %ptr_int %out_final %c1
  %ptr_fin_2 = OpAccessChain %ptr_int %out_final %c2
  %ptr_fin_3 = OpAccessChain %ptr_int %out_final %c3

               OpStore %ptr_fin_0 %final0
               OpStore %ptr_fin_1 %final1
               OpStore %ptr_fin_2 %final2
               OpStore %ptr_fin_3 %final3

               OpReturn
               OpFunctionEnd

Comment (tonghaining) [reply dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsAtomic.java]: Added

File: dartagnan/src/test/java/com/dat3m/dartagnan/spirv/SpirvChecksTest.java [hunk 4]
```diff
@@ -87,6 +87,7 @@ public static Iterable<Object[]> data() throws IOException {
                 {"gpu-verify/atomics/atomic_read_race.spv.dis", 1, PASS},
                 {"gpu-verify/atomics/counter.spv.dis", 1, PASS},
                 // {"gpu-verify/atomics/definitions_atom_int.spv.dis", 1, PASS},
+                 {"gpu-verify/atomics/definitions_atom_int1.spv.dis", 1, PASS},
```
[the remark this answers is not in the document]
Comment (tonghaining) [reply dartagnan/src/test/java/com/dat3m/dartagnan/spirv/SpirvChecksTest.java]: okay

File: dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsLogical.java [hunk 5]
```diff
@@ -45,6 +45,29 @@ public Event visitOpLogicalNot(SpirvParser.OpLogicalNotContext ctx) {
...
+    @Override
+    public Event visitOpSelect(SpirvParser.OpSelectContext ctx) {
+        String id = ctx.idResult().getText();
+        Expression cond = getOperandBoolean(id, ctx.condition().getText());
+        Expression op1 = builder.getExpression(ctx.object1().getText());
+        Expression op2 = builder.getExpression(ctx.object2().getText());
+        Type type = builder.getType(ctx.idResultType().getText());
+        Register register = builder.addRegister(id, ctx.idResultType().getText());
+        if (!op1.getType().equals(type) || !op2.getType().equals(type)) {
+            throw new ParsingException("Illegal definition for '%s', " +
```
[the remark this answers is not in the document]
Comment (tonghaining) [reply dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsLogical.java:57]: Updated

File: dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsAtomic.java [hunk 6]
```diff
@@ -5,6 +5,8 @@
 import com.dat3m.dartagnan.expression.ExpressionFactory;
 import com.dat3m.dartagnan.expression.Type;
 import com.dat3m.dartagnan.expression.integers.IntBinaryOp;
+import com.dat3m.dartagnan.expression.integers.IntCmpOp;
```
[the remark this answers is not in the document]
Comment (tonghaining) [reply dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsAtomic.java:8]: Removed

File: dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsLogical.java [hunk 7]
```diff
@@ -45,6 +45,29 @@ public Event visitOpLogicalNot(SpirvParser.OpLogicalNotContext ctx) {
...
+                    "operands have different types: '%s' is '%s' and '%s' is '%s'",
+                    id, ctx.object1().getText(), op1.getType(), ctx.object2().getText(), op2.getType());
+        }
+        if (op1.getType() instanceof IntegerType) {
+            return builder.addEvent(new Local(register, EXPR_FACTORY.makeITE(cond, op1, op2)));
+        }
+        if (op1.getType() instanceof ArrayType arrayType && arrayType.getElementType() instanceof BooleanType) {
```
[the remark this answers is not in the document]
Comment (natgavrilenko) [reply dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsLogical.java:64]: I see the array is still here, no?

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorSpirvVulkan.java [hunk 8]
```diff
@@ -108,6 +110,29 @@ public List<Event> visitSpirvCmpXchg(SpirvCmpXchg e) {
...
     }
 
+    @Override
+    public List<Event> visitSpirvRMWExtremum(SpirvRmwExtremum e) {
+        e.addTags(Tag.Spirv.MEM_VISIBLE, Tag.Spirv.MEM_AVAILABLE, Tag.Spirv.MEM_NON_PRIVATE);
+        String mo = moToVulkanTag(Tag.Spirv.getMoTag(e.getTags()));
+        Register resultRegister = e.getResultRegister();
+        Register dummy = e.getFunction().newRegister(resultRegister.getType());
+        Load load = newRMWLoadWithMo(dummy, e.getAddress(), Tag.Vulkan.loadMO(mo));
+        Expression originalValue = ExpressionFactory.getInstance().makeIntegerCast(dummy, TypeFactory.getInstance().getArchType(), true);
```
Comment (natgavrilenko) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorSpirvVulkan.java:120]: Why do we need all these casts? Can we just use the original expressions?

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveUnusedMemory.java [hunk 9]
```diff
@@ -3,9 +3,12 @@
 import com.dat3m.dartagnan.expression.Expression;
 import com.dat3m.dartagnan.expression.processing.ExpressionInspector;
 import com.dat3m.dartagnan.program.Program;
+import com.dat3m.dartagnan.program.Register;
```
Comment (natgavrilenko) [inline dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveUnusedMemory.java:6]: unused

Comment (tonghaining) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/RemoveUnusedMemory.java]: removed

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorSpirvVulkan.java (hunk 8 above)
Comment (tonghaining) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorSpirvVulkan.java]: The original expressions work fine here, casts are dropped.

File: dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsLogical.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (tonghaining) [reply dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsLogical.java]: Ops, sorry, should be removed now.

File: dartagnan/src/test/java/com/dat3m/dartagnan/spirv/SpirvChecksTest.java [hunk 10]
```diff
@@ -86,7 +86,7 @@ public static Iterable<Object[]> data() throws IOException {
                 // {"gpu-verify/alignement/race_location.spv.dis", 1, PASS},
                 {"gpu-verify/atomics/atomic_read_race.spv.dis", 1, PASS},
                 {"gpu-verify/atomics/counter.spv.dis", 1, PASS},
-                // {"gpu-verify/atomics/definitions_atom_int.spv.dis", 1, PASS},
+                 {"gpu-verify/atomics/definitions_atom_int.spv.dis", 1, PASS},
```
Comment (natgavrilenko) [inline dartagnan/src/test/java/com/dat3m/dartagnan/spirv/SpirvChecksTest.java:89]: Formatting (extra space), Please also enable this case for SpirvRacesTest

File: dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsAtomic.java [hunk 11]
```diff
@@ -29,7 +30,7 @@ public VisitorOpsAtomic(ProgramBuilderSpv builder) {
     @Override
     public Event visitOpAtomicLoad(SpirvParser.OpAtomicLoadContext ctx) {
         Register register = builder.addRegister(ctx.idResult().getText(), ctx.idResultType().getText());
-        Expression ptr = getPointer(ctx.pointer().getText());
```
Comment (natgavrilenko) [inline dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsAtomic.java:32]: Why not getPointer anymore?

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/event/lang/spirv/SpirvRmwExtremum.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
+package com.dat3m.dartagnan.program.event.lang.spirv;
+
+import com.dat3m.dartagnan.expression.Expression;
+import com.dat3m.dartagnan.expression.integers.IntCmpOp;
+import com.dat3m.dartagnan.program.Register;
+import com.dat3m.dartagnan.program.event.EventVisitor;
+import com.dat3m.dartagnan.program.event.common.RMWOpResultBase;
```
[the remark this answers is not in the document]
Comment (tonghaining) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/event/lang/spirv/SpirvRmwExtremum.java]: removed.

File: dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorSpirvVulkan.java [hunk 13]
```diff
@@ -1,15 +1,17 @@
 package com.dat3m.dartagnan.program.processing.compilation;
 
+import com.dat3m.dartagnan.expression.Expression;
```
[the remark this answers is not in the document]
Comment (tonghaining) [reply dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/compilation/VisitorSpirvVulkan.java:3]: Removed

File: dartagnan/src/test/java/com/dat3m/dartagnan/spirv/SpirvChecksTest.java (hunk 10 above)
Comment (tonghaining) [reply dartagnan/src/test/java/com/dat3m/dartagnan/spirv/SpirvChecksTest.java]: Enabled

File: dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsAtomic.java (hunk 11 above)
Comment (tonghaining) [reply dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsAtomic.java]: changed back

File: dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsAtomic.java [hunk 14]
```diff
@@ -214,7 +262,11 @@ public Set<String> getSupportedOps() {
                 "OpAtomicIDecrement",
                 "OpAtomicAnd",
                 "OpAtomicOr",
-                "OpAtomicXor"
+                "OpAtomicXor",
+                "OpAtomicSMax",
+                "OpAtomicSMin",
+                "OpAtomicUMax",
```
Comment (natgavrilenko) [inline dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/spirv/VisitorOpsAtomic.java:268]: Shouldn't we drop umax/smax for now, until we have a proper support for them?

**Pull request merged**